design/scheduler.md

```diff
@@ -0,0 +1,163 @@
+that provides a human-readable explanation of what an unsuccessful result from
+`Check` means. `Explain` is used to produce a message inside the task that
+explains what is preventing it from being scheduled.
+
+## Scheduling algorithm
+
+The current scheduling algorithm works by building a tree of nodes which is
+specific to the service, and attempting to equalize the total number of tasks
+of this service below the branches of the tree at each level. This is done
+subject to constraints, so a node that, for example, doesn't have enough
+resources to accommodate more tasks, will end up with fewer than its peers.
+
+By default, this tree has only one level, and contains all suitable nodes at
+that level. When [placement preferences](topology.md) are specified, the tree
+can be customized to equalize the number of tasks across specific sets of
+nodes.
+
+While the primary scheduling criterion is the number of tasks from the same
+service on the node, the total number of tasks on the node is used as a
+tiebreaker. The first priority is spreading tasks from each service over as many
+nodes as possible, as evenly as possible, but when there's a choice between
+suitable nodes for the next task, preference is given to the node with the
+fewest total tasks. Note that this doesn't take into consideration things like
+resource reservations and actual resource usage, so this is an area where there
+may be a lot of room for future improvement.
+
+## Batching
+
+The most expensive part of scheduling is building the tree described above. This
+is `O(# nodes)`. If there were `n` nodes and `t` tasks to be scheduled,
+scheduling those tasks independently would have `O(n*t)` runtime. We want to do
+better than this.
+
+A key optimization is that many tasks are effectively identical for the
+scheduler's purposes, being generated by the same service. For example, a
+replicated service with 1000 replicas will cause 1000 tasks to be created, but
+those tasks can be views as equivalent from the scheduler's perspective (until
+they are assigned nodes).
+
+If the scheduler can identify a group of identical tasks, it can build a single
+tree to be shared between them, instead of building a separate tree for each
+one. It does this using the combination of service ID and `SpecVersion`. If
+some number of tasks have the same service ID and `SpecVersion`, they get
+scheduled as a batch using a single tree.
+
+A slight complication with this is that the scheduler receives tasks one by one,
+over a watch channel. If it processed each task immediately, there would be no
+opportunities to group tasks and avoid redundant work. To solve this problem,
+the scheduler waits up to 50 ms after receiving a task, in hopes of receiving of
+another identical task. The total latency associated with this batching is
+limited to one second.
+
+## Building and using the tree
+
+The tree starts out as a tree of max-heaps containing node objects. The primary
```

Comment:
Non-blocking:  I personally found the tree of max heaps a little confusing since in the no-placement-preference case, it seems like just a 1-node tree?  Would it be useful to have an example?  Something like (and please correct me if my understanding of how this works is incorrect):

```
| node | S1 tasks | S2 tasks |       labels            |  
|------|----------|----------|-------------------------|
|   1  |     1    |     1    | engine.labels.os=ubuntu |
|   2  |     1    |     0    | engine.labels.os=ubuntu |
|   3  |     0    |     1    | engine.labels.os=centos |
```
And we want to add one more S2 task (because it was scaled up)

If there are no placement preferences, the tree (one max-heap of nodes) looks like this:

```
               node1
             /       \
          node2     node3
```

Because `node1` has 2 tasks total, and 1 S2 task already. `node2` only has 1 task, and no S2 tasks,
and node3 also only has 1 task but already has an S2 task.


If there are placement preferences, say to spread over `engine.label.ubuntu`:

```
          [root]
            / \
    "ubuntu"   "centos"
    max heap:   max heap:
      node1       node3
        |
      node2
```